### PR TITLE
Creating NodePort test for Windows

### LINKED ISSUE #73327: Need NodePort test for mixed-OS clusters
We need a test case that will verify that NodePort works on all nodes in the cluster for a Windows application.

Here's an example manual test:

Run `kubectl create -f` with this yaml

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: iis-2019
  labels:
    app: iis-2019
spec:
  replicas: 1
  template:
    metadata:
      name: iis-2019
      labels:
        app: iis-2019
    spec:
      containers:
      - name: iis
        image: mcr.microsoft.com/windows/servercore/iis:windowsservercore-ltsc2019
        resources:
          limits:
            cpu: 1
            memory: 800m
          requests:
            cpu: .1
            memory: 300m
        ports:
          - containerPort: 80
      nodeSelector:
        "beta.kubernetes.io/os": windows
  selector:
    matchLabels:
      app: iis-2019
---
apiVersion: v1
kind: Service
metadata:
  name: iis
spec:
  type: NodePort
  ports:
  - protocol: TCP
    port: 80
  selector:
    app: iis-2019
```

Once the pod is running, get the nodeport and make sure it's accessible on each node IP for that port.

```
$ kubectl get svc
NAME         TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)        AGE
iis          NodePort    10.0.88.64   <none>        80:30220/TCP   10m
kubernetes   ClusterIP   10.0.0.1     <none>        443/TCP        10d
azureuser@k8s-master-30163644-0:~$ kubectl get node -o wide
NAME                        STATUS   ROLES   AGE   VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE                    KERNEL-VERSION   CONTAINER-RUNTIME
3016k8s010                  Ready    agent   10d   v1.13.2   10.240.0.35   <none>        Windows Server Datacenter   10.0.17763.253
                            docker://18.9.0
3016k8s011                  Ready   agent   10d   v1.13.2   10.240.0.96   <none>   Windows Server Datacenter   10.0.17763.253
                            docker://18.9.0
k8s-linuxpool1-30163644-0   Ready   agent    10d   v1.13.2   10.240.0.4     <none>   Ubuntu 16.04.5 LTS   4.15.0-1036-azure   docker://3.0.1
k8s-linuxpool1-30163644-1   Ready   agent    10d   v1.13.2   10.240.0.127   <none>   Ubuntu 16.04.5 LTS   4.15.0-1036-azure   docker://3.0.1
k8s-master-30163644-0       Ready   master   10d   v1.13.2   10.255.255.5   <none>   Ubuntu 16.04.5 LTS   4.15.0-1036-azure   docker://3.0.1
azureuser@k8s-master-30163644-0:~$ curl -s -o /dev/null -w "%{http_code}" http://10.240.0.35:30220
200azureuser@k8s-master-30163644-0:~$ curl -s -o /dev/null -w "%{http_code}" http://10.240.0.96:30220
200azureuser@k8s-master-30163644-0:~$ curl -s -o /dev/null -w "%{http_code}" http://10.240.0.4:30220
200azureuser@k8s-master-30163644-0:~$ curl -s -o /dev/null -w "%{http_code}" http://10.240.0.127:30220
200
```

/good-first-issue
/kind test
/sig windows

### PR DESCRIPTION
**Pull request by brusMX**
Co-authored-by: Muzz Imam <muzzz.imam@gmail.com>

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>

/sig windows
/sig testing

**What this PR does / why we need it**:

Tests nodeport for Windows

Fixes  #73327
**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

Re doing PR from https://github.com/kubernetes/kubernetes/pull/74684
cc @michmike @neolit123 

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Review by neolit123 — APPROVED** [review]
/lgtm
/ok-to-test
/priority important-soon
/area test
/kind feature
/sig windows
/sig testing
/assign @michmike

File: test/e2e/windows/service.go
Comment (michmike) [inline test/e2e/windows/service.go]: in windows, it is not "./curl-output.txt" but rather ".\curl-output.txt"
both work, but semantically the other one is the accurate one

**Review by michmike — APPROVED** [review]
/lgtm

**Comment by adelina-t** [comment]
/hold

Tests actually fail for a reason, no need to retest.

File: test/e2e/windows/service.go
Comment (adelina-t) [inline test/e2e/windows/service.go]: We shouldn't be hardcoding image names here. All our images are imported from here: "k8s.io/kubernetes/test/utils/image"

( see other tests in this package for reference )

**Review by adelina-t — COMMENTED** [review]
For sig-windows tests we have to add this to BeforeEach. 

```
       ginkgo.BeforeEach(func() {
		framework.SkipUnlessNodeOSDistroIs("windows")
	})
```

It instructs ginkgo to run this test only when --node-os-distro=windows is passed. Without this in BeforeEach, this tests always runs, including on linux clusters. This is why tests on gce fail, they try running this tests and will never manage to spawn a container with this windows image.

**Comment by bclau** [comment]
Test is also being run on Linux, which it shouldn't.

File: test/e2e/windows/service.go
Comment (brusMX) [reply test/e2e/windows/service.go]: This has been discussed in all the other tests. 
https://onsi.github.io/ginkgo/#avoiding-dot-imports

According to their documentation, we should keep it with dot.

Comment (neolit123) [reply test/e2e/windows/service.go]: @brusMX please remove the dot and reference the `ginko` exported functions like so:
https://github.com/kubernetes/kubernetes/commit/e2f42845d198b02f6472ad7e9831c9fac05dc6a8#diff-05961b7df889d391895155333aa9f488

Comment (brusMX) [reply test/e2e/windows/service.go]: done

Comment (neolit123) [reply test/e2e/windows/service.go]: their recommendation is not a good one. it's a bad programming practice.
the alternative is to keep the dot but add back the `test/e2e/windows` entry in `hack/.golint_failures`

File: test/e2e/windows/service.go
Comment (brusMX) [reply test/e2e/windows/service.go]: Thanks @adelina-t massive (more elegant) re-factoring coming up because of your comment

**Comment by brusMX** [comment]
It should be good now for the /lgtm @adelina-t @PatrickLang @neolit123

**Review by neolit123 — COMMENTED** [review]
/lgtm

**Comment by michmike** [comment]
I am about to remove the hold now.
@adelina-t  i hope that's ok. tests are passing now

**Comment by neolit123** [comment]
@brusMX PR needs rebase also there seem to failures.
you can catch some of these locally without pushing a PR:

```
./hack/vefity-bazel.sh
./hack/verify-gofmt.sh
./hack/verify-spelling.sh
# golint ... the package you are changing
# go vet ... the package you are changing

# check if the test suite builds
bazel build //test/e2e:e2e.test

etc
```

**Review by adelina-t — COMMENTED** [review]
This commit doesn't seem right to me. Why would we merge master branch into this one? Why not just simple rebase?

**Comment by brusMX** [comment]
@adelina-t Well it seems like it was adding only one line at the BUILD file. So I thought it would be more convenient.

**Comment by neolit123** [comment]
> Merge branch 'master' into win-e2e-nodeport

we don't accept merge commits like this one.
please just squash all changes into a single commit.

**Comment by adelina-t** [comment]
> @adelina-t Well it seems like it was adding only one line at the BUILD file. So I thought it would be more convenient.

@brusMX It's bad practice to merge master into other branches. Secondly, it has tons of other changes, the BUILD one is just the only one that github interface is showing because it's resonably sized. 
screenshot 2019-03-06 at 20 40 17

**Comment by brusMX** [comment]
Sorry about that @adelina-t

**Comment by brusMX** [comment]
/lgtm rebased and ready to be merged 
cc @PatrickLang @adelina-t @michmike @neolit123

**Review by michmike — APPROVED** [review]

**Comment by michmike** [comment]
/lgtm
@adelina-t  once you approve, you can remove the hold

**Comment by bclau** [comment]
Just an fyi, the test waits at most 2 minutes for the pod to be Running. In the several runs I've done of this test, the pod becomes Running 1:55 minutes, which is too close to the threshold. Which can be even worse when running tests in parallel. The test should wait 5 minutes for the Pod to become Running, like most other tests.

**Comment by adelina-t** [comment]
> Just an fyi, the test waits at most 2 minutes for the pod to be Running. In the several runs I've done of this test, the pod becomes Running 1:30-1:55 minutes, which is too close to the threshold. Which can be even worse when running tests in parallel. The test should wait 5 minutes for the Pod to become Running, like most other tests.

Sadly, 2 minute timeout is hardcoded in jig :( https://github.com/kubernetes/kubernetes/blob/b67126e350fdba6d89a0ee64310cac7db14e33d3/test/e2e/framework/service_util.go#L755

**Comment by adelina-t** [comment]
Ok, this looks good to me. Tested it a bunch of times, works as expected.

Thank you @brusMX :)

/hold cancel
/approve

**Pull request merged**
